xla/service/cpu/onednn_contraction_rewriter.cc

```diff
@@ -464,6 +494,74 @@ class OneDnnMatMulRewriteVisitor : public DfsHloRewriteVisitor {
     return absl::OkStatus();
   }
 
+  absl::Status HandleConvolution(HloInstruction* conv) override {
+    auto pattern = match::Op(&conv).WithOpcode(HloOpcode::kConvolution);
+    if (!Match(conv, pattern)) return absl::OkStatus();
+    if (!OneDnnContractionRewriter::ShouldRewriteConv(conv)) {
+      return absl::OkStatus();
+    }
+
+    const Shape& conv_shape = conv->shape();
+    auto dims = conv->window().dimensions().size();
+    const ConvolutionDimensionNumbers& conv_dims =
+        conv->convolution_dimension_numbers();
+
+    BackendConfig backend_config;
+    OneDnnConvolutionConfig* conv_config =
+        backend_config.mutable_onednn_conv_config();
+
+    conv_config->set_dims(conv_shape.rank());
+    conv_config->set_feature_groups(conv->feature_group_count());
+    conv_config->mutable_input()->mutable_data()->set_batch_dim(
+        conv_dims.input_batch_dimension());
+    conv_config->mutable_kernel()->mutable_filter()->set_input_feature_dim(
+        conv_dims.kernel_input_feature_dimension());
+    conv_config->mutable_output()->mutable_data()->set_batch_dim(
+        conv_dims.output_batch_dimension());
+    conv_config->mutable_input()->mutable_data()->set_feature_dim(
+        conv_dims.input_feature_dimension());
+    conv_config->mutable_kernel()->mutable_filter()->set_output_feature_dim(
+        conv_dims.kernel_output_feature_dimension());
+    conv_config->mutable_output()->mutable_data()->set_feature_dim(
+        conv_dims.output_feature_dimension());
+
+    const Shape& output_shape = conv->shape();
+
+    for (auto it = conv->window().dimensions().begin();
+         it != conv->window().dimensions().end(); it++) {
+      if ((*it).padding_low() < 0 || (*it).padding_high() < 0 ||
+          (*it).stride() < 0) {
+        return absl::OkStatus();
+      }
+      conv_config->mutable_window()->add_pad_left((*it).padding_low() + 1);
+      conv_config->mutable_window()->add_pad_right((*it).padding_high() + 1);
+      conv_config->mutable_window()->add_strides((*it).stride() + 1);
+      conv_config->mutable_window()->add_window_dilations(
+          (*it).window_dilation() + 1);
+      if ((*it).base_dilation() != 1 || (*it).window_reversal()) {
+        return absl::OkStatus();
+      }
+    }
+
+    for (int i = 0; i < dims; i++) {
+      conv_config->mutable_input()->mutable_data()->add_spatial_dims(
+          conv_dims.input_spatial_dimensions()[i] + 1);
```

Comment:
Why do we `+ 1` every sizes here (and above)? Please document the reason in the code.

Reply:
We add 1 to change the input subspace of uint repeated fields from whole numbers ```[0, ...)``` to natural numbers ```[1, ...)```. If we try to store ```0``` as one of the values in the field, then all the values following the ```0``` are lost.

Reply:
Thank you!